src/ARTED/common/hpsi.f90

```diff
@@ -150,34 +155,54 @@ subroutine subtraction(Vloc,tpsi,htpsi,ttpsi)
       end do
     end subroutine
 
-    subroutine pseudo_pt(ik,tpsi,htpsi)
-#ifdef ARTED_STENCIL_PADDING
-      use opt_variables, only: zJxyz => zKxyz
-#else
-      use opt_variables, only: zJxyz
-#endif
+    !Calculating nonlocal part
+    subroutine pseudo_pt(ik,idx_proj,zproj,tpsi,htpsi)
+      use opt_variables, only: NP,NPI,pseudo_start_idx
+      use global_variables, only: NI,Nps
       implicit none
       integer,    intent(in)  :: ik
+      integer,    intent(in)  :: idx_proj(NPI)
+      complex(8), intent(in)  :: zproj(Nps,Nlma)
       complex(8), intent(in)  :: tpsi(0:PNL-1)
       complex(8), intent(out) :: htpsi(0:PNL-1)
-      integer    :: ilma,ia,j,i
+
+      integer    :: ia,i,j,ip,ioffset
       complex(8) :: uVpsi
 
-      !Calculating nonlocal part
-      do ilma=1,Nlma
-        ia=a_tbl(ilma)
-        uVpsi=0.d0
+      complex(8) :: pseudo(NPI)
+      complex(8) :: dpseudo(NPI)
+
+      dpseudo = cmplx(0.d0)
+
+      ! gather (load) pseudo potential point
+      do i=1,NPI
+        pseudo(i) = tpsi(idx_proj(i))
+      end do
+
+      do ia=1,NI
+      do ip=1,NP
+        i = NP*(ia-1) + ip
+
+        ! summarize vector
+        uVpsi   = 0.d0
+        ioffset = pseudo_start_idx(ia)
         do j=1,Mps(ia)
-          i=zJxyz(j,ia)
-          uVpsi=uVpsi+conjg(zproj(j,ilma,ik))*tpsi(i)
+          uVpsi = uVpsi + conjg(zproj(j,i)) * pseudo(ioffset+j)
         end do
-        uVpsi=uVpsi*Hxyz*iuV(ilma)
-!dir$ ivdep
+        uVpsi = uVpsi * Hxyz * iuV(i)
+
+        ! apply vector
+        ioffset = pseudo_start_idx(ia)
         do j=1,Mps(ia)
-          i=zJxyz(j,ia)
-          htpsi(i)=htpsi(i)+zproj(j,ilma,ik)*uVpsi
+          dpseudo(ioffset+j) = dpseudo(ioffset+j) + zproj(j,i) * uVpsi
         end do
       end do
+      end do
+
+      ! scatter (store) pseudo potential point
+      do i=1,NPI
```

Comment:
The use of the full array `idx_proj` may induce segmentation fault. Please have a look at the above comment.